### DISCUSSION TITLE
AdcParam for ESP32

### DISCUSSION THREAD

**Discussion by Sayed-afridy** [arendst/Tasmota#19409] [Q&A]
i have connected ASC712 with the ESP32 . by selecting CT Clamp as adc input . but i didnt find a proper way to calibrate this to work.
on the otherhand Nodemcu works great with this AdcParam7,0,1140,0.22,0.060 . but the esp32 is not working with this

**Comment by sfromis** [comment]
Two things....

1) Your `AdcParam7,0,1140,220,0,060` looks wrong, as the docs say 4 parameters not 5. 
And the docs example has `AdcParam 7,0,2146,0.230,0.060`
Note the last parameter is `0.060` with a decimal point, where you post about using a comma instead.

2) Since the original ESP32 has ADC values 0..4095 instead of 0..1023 (which the setup seems to be based on), I'd speculatively try multiplying (or dividing) the 2nd parameter (where you have 2146) by 4.

Note: I have zero experience with ASC712 or CT power measurements, and am not at all confident of the above speculative idea.

**Reply by barbudor** [reply]
You still haven't said which model of ACS712 you are using. Without that information, there is no way to suggest any close to proper values.

**Reply by Sayed-afridy** [reply]
I am using ACS712 20A and  i have used AdcParam7,0,1140,0.22,0.060 . for nodemcu it works great but for esp32 its not working

**Reply by barbudor** [reply]
An ACS712-20 has a A-to-V ratio of 0.1mV/A and an output centered on 2.5V for 0A
For a 5A(rms) current,  the peak-to-peak output of the ACS712-20A should be 1.414V which translate on the ADC as 438 on ESP8266 which ADC range is 0..1023 for 0..3.3V
The 2nd parameter should be (100000 * 5) / 438 = 1141.55 so your value of 1140 is good

On an ESP32, where the ADC range is 0..4095 for 0.3.3V the ADC measure would be 1755
So the 2nd parameter should be (100000 * 5) / 1755 = 284.88 so a value of 285 would be ok

NOW considering that the ADC input is limited to 3.3V, you cannot measure a current over 5.6A
After that you will start to saturate the upper part of the sin wave and your measures will start to be wrong.

The solution on ESP8266 is to add a resistor in serie so the internal resistor divider provide an input range of up to 4.5V
If you add a 150Kohms in serie in the input, added to the existing 220K/100K divider your range would be 4.7V
For 14Arms, peak-to-peak output of the ACS712 would be 3.96V as your range is 4.7V over 0..1023 the ADC would provide 862
So the 2nd parameter would have to be (100000 * 14) / 862 = 1624

On ESP32 it is slighly different because there is no existing voltage divider. So you need to add one to bring hte range from 3.3V to 4.7V
You can do that with a 150Kohms and and 330kohms which would provide a range of 4.8V
Once again, for 14Arms, peak-to-peak output of the ACS712 would be 3.96V as your range is 4.8V over 0..4095 the ADC would provide 3378
So the 2nd parameter would have to be (100000 * 14) / 3378 = 414

It could probably easier to understand with some diagrams but don't forget the ACS712 provides a voltage that is a conversion of the instantaneous current. So it provides an AC signal to Tasmota which measures the peak-to-peak, and convert it to a current using the 2nd parameter. So you have to take into account the conversion between peak and RMS.

**Reply by Sayed-afridy** [reply]
with no load it provides adc value of 103 . and for ct power it shows around 280w-380w (its not stable) in no load operation with the value of AdcParam7,0,285,0.22,0,060 . how to calibrate this in a way so that it will show 0w in no load mode

**Reply by barbudor** [reply]
> with no load it provides adc value of 103

Which ESP ?
Which mode of ADC ?

When having no load, you should have 2.5V at the output of the ACS712.
With GPIO in "ADC input" mode, you should have a reading of around 775 on nodeMCU/mini-D1 ESP8266, or around 3100 on an ESP32

There is no way to calibrate the 0, only the conversion ratio.
This is why the 4th parameter was introduce : to provide a threshold under which power is assumed 0
